table/block_based/block_based_table_reader.cc

```diff
@@ -2979,26 +2979,41 @@ Status BlockBasedTable::CreateIndexReader(
   }
 }
 
-uint64_t BlockBasedTable::ApproximateOffsetOf(
-    const InternalIteratorBase<IndexValue>& index_iter) const {
-  uint64_t result = 0;
+uint64_t BlockBasedTable::ApproximateDataOffsetOf(
+    const InternalIteratorBase<IndexValue>& index_iter,
+    uint64_t data_size) const {
   if (index_iter.Valid()) {
     BlockHandle handle = index_iter.value().handle;
-    result = handle.offset();
+    return handle.offset();
   } else {
-    // The iterator is past the last key in the file. If table_properties is not
-    // available, approximate the offset by returning the offset of the
-    // metaindex block (which is right near the end of the file).
-    if (rep_->table_properties) {
-      result = rep_->table_properties->data_size;
-    }
-    // table_properties is not present in the table.
-    if (result == 0) {
-      result = rep_->footer.metaindex_handle().offset();
-    }
+    // The iterator is past the last key in the file.
+    return data_size;
+  }
+}
+
+uint64_t BlockBasedTable::GetApproximateDataSize() {
+  // First try table properties
+  if (rep_->table_properties) {
+    return rep_->table_properties->data_size;
   }
+  // Then try blocks right after data
+  if (rep_->filter_handle != BlockHandle()) {
+    assert(!rep_->filter_handle.IsNull());
+    return rep_->filter_handle.offset();
+  }
+  // Index handle not readily available, so...
+  if (rep_->compression_dict_handle != BlockHandle()) {
+    assert(!rep_->compression_dict_handle.IsNull());
+    return rep_->compression_dict_handle.offset();
+  }
+  // Hmm, use rough estimate from footer
+  return rep_->footer.metaindex_handle().offset();
```

Comment:
rep_->table_properties is only null if the file was generated by release some early 2.x releases or earlier. This is a rare case. My vote would be to simplify the case. We don't have to try filter and directly go to metaindex. And I believe dictionary compression was not available for those early versions.